src/sentry/search/events/builder.py

```diff
@@ -0,0 +1,272 @@
+from datetime import datetime
+from typing import Any, Callable, List, Mapping, Optional, Sequence, Union
+
+from parsimonious.exceptions import ParseError
+from snuba_sdk.column import Column
+from snuba_sdk.conditions import BooleanCondition, Condition, Op, Or
+from snuba_sdk.entity import Entity
+from snuba_sdk.expressions import Limit
+from snuba_sdk.function import CurriedFunction, Function
+from snuba_sdk.orderby import Direction, OrderBy
+from snuba_sdk.query import Query
+
+from sentry.api.event_search import SearchFilter, SearchKey, SearchValue, parse_search_query
+from sentry.exceptions import InvalidSearchQuery
+from sentry.search.events.constants import (
+    EQUALITY_OPERATORS,
+    NO_CONVERSION_FIELDS,
+    TAG_KEY_RE,
+    VALID_FIELD_PATTERN,
+)
+from sentry.search.events.fields import FIELD_ALIASES, is_function
+from sentry.utils.snuba import Dataset, resolve_column
+
+# TODO: this should be a TypedDict instead
+ParamsType = Mapping[str, Union[List[int], int, str, datetime]]
+AggregateType = Union[Function, CurriedFunction]
+WhereType = Union[Condition, BooleanCondition]
+SelectType = Union[Column, CurriedFunction, Function]
+
+OPERATOR_MAP: Mapping[str, Op] = {
+    "=": Op.EQ,
+    "!=": Op.NEQ,
+    "IN": Op.IN,
+    "NOT IN": Op.NOT_IN,
+}
+
+
+class QueryBuilder:
+    """ Builds a snql query """
+
+    # Allow list of fields that are handled by this class Once we reach a
+    # certain threshold of fields handled should turn this into a denylist
+    # use public facing field/function names for this list
+    field_allowlist = {
+        "user.email",
+        "release",
+        "environment",
+    }
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        params: ParamsType,
+        query: Optional[str] = None,
+        selected_columns: Optional[List[str]] = None,
+        orderby: Optional[List[str]] = None,
+        limit: int = 50,
+    ):
+        self.aggregates: List[AggregateType] = []
+        self.columns: List[SelectType] = []
+        self.where: List[WhereType] = []
+
+        self.params = params
+        self.dataset = dataset
```

Comment:
I wonder if it would be cleaner to have a `self.query: Query` that you call the `set_*` functions on instead of having `self.where`, `self.limit` etc. I'm not saying you need to change this, but something to consider.

Reply:
👍 Yeah I had considered that too, but since Query is immutable so I figured this way its easier to edit the clauses at will and create one Query instance at the very end.